source/common/upstream/load_balancer_impl.cc

```diff
@@ -670,10 +795,13 @@ uint32_t ZoneAwareLoadBalancerBase::tryChooseLocalLocalityHosts(const HostSet& h
 
   // This potentially can be optimized to be O(log(N)) where N is the number of localities.
   // Linear scan should be faster for smaller N, in most of the scenarios N will be small.
-  // TODO(htuch): is there a bug here when threshold == 0? Seems like we pick
-  // local locality in that situation. Probably should start iterating at 1.
+  //
+  // Bucket 1: [0, state.residual_capacity_[0] - 1]
+  // Bucket 2: [state.residual_capacity_[0], state.residual_capacity_[1] - 1]
+  // ...
+  // Bucket N: [state.residual_capacity_[N-2], state.residual_capacity_[N-1] - 1]
   int i = 0;
-  while (threshold > state.residual_capacity_[i]) {
+  while (threshold >= state.residual_capacity_[i]) {
     i++;
   }
```

Comment:
fix off-by-one bug that doesn't matter much because our precision loss from using integers is higher than this error would be.